agent/logs.go

```diff
@@ -118,15 +121,30 @@ func (l *logSender) sendLoop(ctx context.Context, dest logDest) error {
 			return nil
 		}
 		src, q := l.getPendingWorkLocked()
+		logger := l.logger.With(slog.F("log_source_id", src))
 		q.flushRequested = false // clear flag since we're now flushing
 		req := &proto.BatchCreateLogsRequest{
 			LogSourceId: src[:],
-			// when we release the lock, we don't want modifications to the slice to affect us
-			Logs: slices.Clone(q.logs),
+		}
+		o := 0
+		n := 0
+		for n < len(q.logs) {
+			log := q.logs[n]
+			if len(log.Output) > logOutputMaxBytes {
+				logger.Warn(ctx, "dropping log line that exceeds our limit")
+				n++
+				continue
+			}
+			o += len(log.Output) + overheadPerLog
+			if o > logOutputMaxBytes {
+				break
+			}
+			req.Logs = append(req.Logs, log)
+			n++
```

Comment:
I think this change here demonstrates a downside of using PR stacking too much. I don't think it's a good idea to split up an implementation like this, especially a new one (I might be of a different opinion for something much bigger).

Here is a change that introduces more advanced logic in a place where I left a simple suggestion in the previous PR. I think that comment applies here too (as in we don't need a clone of the slice, we can utilize the backing array/append semantics). But based on the comment from my previous PR, this may be easily overlooked due to the change in this one.